executor/foreign_key.go

```diff
@@ -503,3 +519,164 @@ func (fkc FKCheckExec) checkRows(ctx context.Context, sc *stmtctx.StatementConte
 	}
 	return nil
 }
+
+func (b *executorBuilder) buildTblID2FKCascadeExecs(tblID2Table map[int64]table.Table, tblID2FKCascades map[int64][]*plannercore.FKCascade) (map[int64][]*FKCascadeExec, error) {
+	fkCascadesMap := make(map[int64][]*FKCascadeExec)
+	for tid, tbl := range tblID2Table {
+		fkCascades, err := b.buildFKCascadeExecs(tbl, tblID2FKCascades[tid])
+		if err != nil {
+			return nil, err
+		}
+		if len(fkCascades) > 0 {
+			fkCascadesMap[tid] = fkCascades
+		}
+	}
+	return fkCascadesMap, nil
+}
+
+func (b *executorBuilder) buildFKCascadeExecs(tbl table.Table, fkCascades []*plannercore.FKCascade) ([]*FKCascadeExec, error) {
+	fkCascadeExecs := make([]*FKCascadeExec, 0, len(fkCascades))
+	for _, fkCascade := range fkCascades {
+		fkCascadeExec, err := b.buildFKCascadeExec(tbl, fkCascade)
+		if err != nil {
+			return nil, err
+		}
+		if fkCascadeExec != nil {
+			fkCascadeExecs = append(fkCascadeExecs, fkCascadeExec)
+		}
+	}
+	return fkCascadeExecs, nil
+}
+
+func (b *executorBuilder) buildFKCascadeExec(tbl table.Table, fkCascade *plannercore.FKCascade) (*FKCascadeExec, error) {
+	colsOffsets, err := getFKColumnsOffsets(tbl.Meta(), fkCascade.ReferredFK.Cols)
+	if err != nil {
+		return nil, err
+	}
+	helper := &fkValueHelper{
+		colsOffsets: colsOffsets,
+		fkValuesSet: set.NewStringSet(),
+	}
+	return &FKCascadeExec{
+		b:             b,
+		fkValueHelper: helper,
+		tp:            fkCascade.Tp,
+		referredFK:    fkCascade.ReferredFK,
+		childTable:    fkCascade.ChildTable.Meta(),
+		fk:            fkCascade.FK,
+	}, nil
+}
+
+func (fkc *FKCascadeExec) onDeleteRow(sc *stmtctx.StatementContext, row []types.Datum) error {
+	vals, err := fkc.fetchFKValuesWithCheck(sc, row)
+	if err != nil || len(vals) == 0 {
+		return err
+	}
+	fkc.fkValues = append(fkc.fkValues, vals)
+	return nil
+}
+
+func (fkc *FKCascadeExec) buildExecutor(ctx context.Context) (Executor, error) {
+	p, err := fkc.buildFKCascadePlan(ctx)
+	if err != nil || p == nil {
+		return nil, err
+	}
+	e := fkc.b.build(p)
+	return e, fkc.b.err
+}
+
+func (fkc *FKCascadeExec) buildFKCascadePlan(ctx context.Context) (plannercore.Plan, error) {
+	if len(fkc.fkValues) == 0 {
+		return nil, nil
+	}
+	var sqlStr string
+	var err error
+	switch fkc.tp {
+	case plannercore.FKCascadeOnDelete:
+		sqlStr, err = GenCascadeDeleteSQL(fkc.referredFK.ChildSchema, fkc.childTable.Name, fkc.fk, fkc.fkValues)
+	}
+	if err != nil || sqlStr == "" {
+		return nil, err
+	}
+
+	sctx := fkc.b.ctx
+	exec, ok := sctx.(sqlexec.RestrictedSQLExecutor)
+	if !ok {
+		return nil, nil
+	}
+	stmtNode, err := exec.ParseWithParams(ctx, sqlStr)
+	if err != nil {
+		return nil, err
+	}
+	err = ResetContextOfStmt(sctx, stmtNode)
```

Comment:
The `ResetContextOfStmt` is called by the execution framework. The function `uildFKCascadePlan` does something the same as the framework does, maybe it's better to reuse the existing functions by making necessary changes to them.

Reply:
It really doesn't quite fit here, so it has been deleted.

Reply:
I have refactored the logic here, PTAL